nautobot/core/apps/__init__.py

```diff
@@ -0,0 +1,246 @@
+from django.apps import AppConfig
+from django.urls import reverse
+from operator import getitem
+from collections import OrderedDict
+
+from nautobot.extras.plugins.utils import import_object
+from nautobot.extras.registry import registry
+from nautobot.utilities.choices import ButtonActionColorChoices, ButtonActionIconChoices
+
+
+registry["nav_menu"] = {"tabs": {}}
+
+
+class NautobotConfig(AppConfig):
+    """
+    Custom AppConfig for Nautobot application.
+
+    Adds functionality to generate the HTML navigation menu using `navigation.py` files from nautbot
+    applications.
+    """
+
+    menu_tabs = "navigation.menu_tabs"
+
+    def ready(self):
+        """
+        Ready function initiates the import application.
+        """
+        menu_items = import_object(f"{self.name}.{self.menu_tabs}")
+        if menu_items is not None:
+            register_menu_items(menu_items)
+
+
+def register_menu_items(tab_list):
+    """
+    Using the imported object a dictionary is either created or updated with objects to create
+    the navbar.
+
+    The dictionary is built from four key objects, NavMenuTab, NavMenuGroup, NavMenuItem and
+    NavMenuButton. The Django template then uses this dictionary to generate the navbar HTML.
+    """
+    for nav_tab in tab_list:
+        if nav_tab.name not in registry["nav_menu"]["tabs"]:
+            registry["nav_menu"]["tabs"][nav_tab.name] = {
+                "weight": nav_tab.weight,
+                "groups": {},
+                "permissions": [],
+            }
+
+        tab_perms = []
+        registry_groups = registry["nav_menu"]["tabs"][nav_tab.name]["groups"]
+        for group in nav_tab.groups:
+            if group.name not in registry_groups:
+                registry_groups[group.name] = {
+                    "weight": group.weight,
+                    "items": {},
+                    "permissions": [],
+                }
+
+            group_perms = []
+            for item in group.items:
+                if item.link not in registry_groups[group.name]["items"]:
+                    registry_groups[group.name]["items"][item.link] = {
+                        "link_text": item.link_text,
+                        "weight": item.weight,
+                        "buttons": {},
+                        "permissions": item.permissions,
+                    }
```

Comment:
Perhaps some error checking along the following lines?

```suggestion
                    registry_groups[group.name]["items"][item.link] = {
                        "link_text": item.link_text,
                        "weight": item.weight,
                        "buttons": {},
                        "permissions": item.permissions,
                    }
                else:
                    for key, new_value in (
                        ("link_text", item.link_text),
                        ("weight", item.weight),
                        ("permissions", item.permissions),
                    ):
                        if registry_groups[group.name]["items"][item.link][key] != new_value:
                            logger.error(
                                "Unable to redefine %s on %s -> %s -> %s from %s to %s",
                                key,
                                nav_tab.name,
                                group.name,
                                item.link,
                                registry_groups[group.name]["items"][item.link][key],
                                new_value,
                            )
```


Similar suggestion at other levels of the menu construction as well - basically, any case where there's a mismatch between the value requested and the value that "wins", we should probably log messages to aid in troubleshooting.